File: solutions/python/easy/p0020_valid_parentheses.py
"""LeetCode 20: Valid Parentheses.

Problem: https://leetcode.com/problems/valid-parentheses/
"""


class Solution:
    def isValid(self, s: str) -> bool:
        """Validate closing brackets against the most recent opening bracket.

        Time: O(n)
        Space: O(n)
        """
        pairs = {")": "(", "]": "[", "}": "{"}
        stack: list[str] = []
        for character in s:
            if character in pairs.values():
                stack.append(character)
            elif not stack or stack.pop() != pairs[character]:
                return False
        return not stack

File: solutions/java/easy/P0020_ValidParentheses.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;

/**
 * LeetCode 20: Valid Parentheses.
 *
 * Problem: https://leetcode.com/problems/valid-parentheses/
 */
public class P0020_ValidParentheses {
    public boolean isValid(String s) {
        // Time: O(n), Space: O(n)
        Map<Character, Character> pairs = new HashMap<>();
        pairs.put('(', ')');
        pairs.put('[', ']');
        pairs.put('{', '}');

        Deque<Character> stack = new ArrayDeque<>();

        for (int i = 0; i < s.length(); i++) {
            char current = s.charAt(i);

            if (pairs.containsKey(current)) {
                stack.push(pairs.get(current));
            } else {
                if (stack.isEmpty() || stack.pop() != current) {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }

    public static void main(String[] args) {
        P0020_ValidParentheses solution = new P0020_ValidParentheses();
        System.out.println(solution.isValid("()"));       // Expected: true
        System.out.println(solution.isValid("()[]{}"));   // Expected: true
        System.out.println(solution.isValid("(]"));       // Expected: false
        System.out.println(solution.isValid("([)]"));     // Expected: false
        System.out.println(solution.isValid("{[]}"));     // Expected: true
    }
}
